### PR TITLE
Provide interface for testing

### PR DESCRIPTION
**Pull request by bhelx**
Providing an interface for the Client struct for testing purposes.

### Example Use

```go

// MockClient will be our fake client that implements recurly.ClientInterface
type MockClient struct {
}

// we need to implement each method in the interface
func (c *MockClient) CreateAccount(request *recurly.AccountCreate) (*recurly.Account, error) {
	// we can setup fake scenarios here depending on the input
        // this scenario is hypothetical but it demonstrates that you can use input based rules, or state on the MockClient struct
        // to control the behavior of the client for your testing scenarios
	if *request.Code == "good" {
                // if the account code in our body is "good" we can just return a mock account
		account := &recurly.Account{
			Id:   "abc123",
			Code: *request.Code,
		}
		return account, nil
	} else {
                // if the account code in our body isn't "good", let's return a mock validation error
		err := &recurly.Error{
			Type:    recurly.ErrorTypeValidation,
			Class:   recurly.ErrorClassClient,
			Message: "Account body not valid",
			Params: []recurly.ErrorParam{
				{
					Property: "code",
					Message:  "Account code is not valid",
				},
			},
		}
		return nil, err
	}
}

// if you don't care about or don't use some of these methods, maybe you can just return nil to satisfy the interface
//func (c *MockClient) GetAccount(accountId string) (*recurly.Account, error) {
//    return nil, nil
//}

// In our application, we can use the interface rather than the recurly.Client object
// In testing this can be injected with MockClient, in production it can use Client
func DoRecurlyThings(c recurly.ClientInterface) {
	request := &recurly.AccountCreate{
		Code: recurly.String("good"),
	}
	acc, _ := c.CreateAccount(request)
	fmt.Printf("Created account with id: %s\n", acc.Id)

	requestBad := &recurly.AccountCreate{
		Code: recurly.String("bad"),
	}
	_, err := c.CreateAccount(requestBad)
	if e, ok := err.(*recurly.Error); ok {
		if e.Type == recurly.ErrorTypeValidation {
			fmt.Printf("Failed validation: %v\n", e)
			fmt.Println("Parameters:")
			for _, v := range e.Params {
				fmt.Printf("%s: %s\n", v.Property, v.Message)
			}
		} else {
			fmt.Printf("Unexpected Recurly error: %v\n", e)
		}
	}
}

func main() {
	client := &MockClient{}
        // you can use MockClient for testing
        // or you can use recurly.Client for production / live calls to the server
	DoRecurlyThings(client)
}
```

Prints:

```txt
Created account with id: abc123
Failed validation: Account body not valid
Parameters:
code: Account code is not valid
```

### CODE REVIEW

File: client_operations.go [hunk 1]
```diff
@@ -200,11 +436,7 @@ func (list *ListAccountsParams) URLParams() []KeyValue {
 func (c *Client) ListAccounts(params *ListAccountsParams) *AccountList {
 	path := "/accounts"
 	path = BuildUrl(path, params)
-	return &AccountList{
-		client:       c,
-		nextPagePath: path,
-		HasMore:      true,
-	}
+	return NewAccountList(c, path)
```
Comment (bhelx) [inline client_operations.go:439]: Decided to make constructors for these to make mocking easier

Comment (bhelx) [reply client_operations.go:439]: The programmer can pass in an HttpCaller here instead of a client.

Comment (bmacnee) [reply client_operations.go:439]: yeah i think this is definitely the right call and it keeps our encapsulation nice as well

**Comment by bhelx** [comment]
> My biggest concern is the effort necessary to fake the ClientInterface because it has so many methods. But this can be overcome using a mock generation tool.

Right, I used the mock generator when testing. It's unfortunate, but at the moment, there is no way to logically separate the interfaces in an automated way. We could look into this in the future though.

**Pull request merged**